include/class.orm.php

```diff
@@ -2585,8 +2585,7 @@ function compileQ(Q $Q, $model, $parens=true) {
         }
         $glue = $Q->ored ? ' OR ' : ' AND ';
         $clause = implode($glue, $filter);
-        if (($Q->negated || $parens) && count($filter) > 1)
-            $clause = '(' . $clause . ')';
+        $clause = '(' . $clause . ')';
```

Comment:
Hey @greezybacon - do you foresee any issues with this change? See the PR conversation for more details.

Thanks,

Reply:
I would suggest this patch for the ORM. There is definitely a bug:
```diff
diff --git a/include/class.orm.php b/include/class.orm.php
index 132986a4..d7e12931 100644
--- a/include/class.orm.php
+++ b/include/class.orm.php
@@ -2615,6 +2615,9 @@ class SqlCompiler {
                     }
                 }
                 $filter[] = $this->compileQ(new Q($criteria), $model,
+                    // New criteria here is joined with AND, so if the outer
+                    // criteria is joined with OR or is negated, then
+                    // parentheses are necessary
                     $Q->ored || $Q->negated);
             }
             // Handle simple field = <value> constraints
@@ -2649,16 +2654,10 @@ class SqlCompiler {
 
     function compileConstraints($where, $model) {
         $constraints = array();
-        $prev = $parens = false;
         foreach ($where as $Q) {
-            if ($prev && !$prev->isCompatibleWith($Q)) {
-                $parens = true;
-                break;
-            }
-            $prev = $Q;
-        }
-        foreach ($where as $Q) {
-            $constraints[] = $this->compileQ($Q, $model, $parens);
+            // Constraints are joined by AND operators, so if they have
+            // internal OR operators, then they need to be parenthesized
+            $constraints[] = $this->compileQ($Q, $model, $Q->ored);
         }
         return $constraints;
     }
```